### ISSUE TITLE
Nullability analysis of tuple equality

### ISSUE THREAD

**Issue by jcouv** [dotnet/roslyn#50980]
The nullability analysis of `(a, b, ...) = (x, y, ...)` should be like `a == x && b == y && ...` while keeping the order of evaluation strictly left to right (`a`, `b`, ..., `x`, `y`, ...).

Spec: https://github.com/dotnet/csharplang/blob/master/proposals/csharp-7.3/tuple-equality.md

FYI @RikkiGibson

**Comment by CyrusNajmabadi** [comment]
tuple equality?  or tuple assignment?

**Comment by jcouv** [comment]
Tuple equality. Corrected the snippet in OP. Thanks :-)

**Comment by RikkiGibson** [comment]
Based on the issue description I thought of a few scenarios off-the-cuff: [SharpLab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBMQGoA+BiAdgVwBtCJhC4ACOXU8gWAChHgB7FwigMwEYAKAAW4AGAPwUISCoNEVgASkYBeAHwVeE2XIoBCRWoLFJEAHQAVFgGUYCAJa4A5rzlyA3BQD07qgA8ADnABjGApcFgoAdwgEXDt7RmY2Dk4AJgFhMQ1pMXklVXUjM0trWKctXX0iQkkDQlcPLzg/QOCICKiYhyA)
```cs
#nullable enable

bool f1(string? a, string? b)
=> (a, b) != (null, a.ToString()); // expect no warning

bool f2(string? a, string? b)
=> (a, a.ToString()) != (null, null); // expect a warning
```

**Comment by jcouv** [comment]
Right.

Some more scenario ideas:
```
// pure null test
void M(string s)
{
   if ((s, 1) == (null, 1))
   {
        s.ToString(); /// should warn because of pure null test
   }
}
```

```
    // MemberNotNullWhen
    if ((this.HasValue, 1) == (true, 1))
    {
         _ = this.Value.ToString(); // should not warn, assuming that `HasValue` is marked with `[MemberNotNullWhen(true, "Value")]`
    }
```

```
   // NotNullWhen
   if ((TryGetValue("key", out var value), 1) == (true, 1))
   {
       value.ToString(); // should not warn, assuming that `TryGetValue` has `NotNullWhen(true)` on the `out` parameter
   }
```

I suspect that getting one scenario to work will just get all the scenarios to work. We need to delegate `VisitTupleEquality` to share the core logic of `AfterLeftChildHasBeenVisited` (and possibly `VisitBinaryOperatorChildren`) which deal with comparisons to `null` and Boolean constants.
